api/storage_service.cc

```diff
@@ -600,6 +602,161 @@ rest_get_token_endpoint(http_context& ctx, sharded<service::storage_service>& ss
         }));
 }
 
+static
+future<json::json_return_type>
+rest_set_client_routes(http_context& ctx, sharded<service::storage_service>& ss, service::raft_group0_client& group0_client, std::unique_ptr<http::request> req) {
+    if (!ss.local().get_feature_service().client_routes) {
+        apilog.warn("set_client_routes: called before the cluster feature was enabled");
+        throw std::runtime_error("set_client_routes requires all nodes to support the CLIENT_ROUTES cluster feature");
+    }
+
+    rapidjson::Document root;
+    auto content = co_await util::read_entire_stream_contiguous(*req->content_stream);
+    root.Parse(content.c_str());
+    if (!root.IsArray()) {
+        throw bad_param_exception("Body must be a JSON array");
+    }
+
+    co_await ss.invoke_on(0, [&group0_client, &root] (service::storage_service& ss) -> future<> {
+        int retries = 10;
+        while (true) {
+            try {
```

Comment:
Better to do all json parsing/validation outside of `group0_client.start_operation`, probably even on the original shard, and do the minimal conversion `const vector<client_route>& -> mutations` under the group0 lock